web/docs/guides/middleware-customization.md

```diff
@@ -0,0 +1,134 @@
+
+### Types
+
+Below are the relevant TS types and the actual definitions of default middleware.
+
+```ts
+export type MiddlewareConfig = Map<string, express.RequestHandler>
+
+export type MiddlewareConfigFn = (middleware: MiddlewareConfig) => MiddlewareConfig
+
+const defaultGlobalMiddleware: MiddlewareConfig = new Map([
+  ['helmet', helmet()],
+  ['cors', cors({ origin: config.allowedCORSOrigins })],
+  ['logger', logger('dev')],
+  ['express.json', express.json()],
+  ['express.urlencoded', express.urlencoded({ extended: false })],
+  ['cookieParser', cookieParser()]
+])
+```
+
+Here is the [definition of config](https://github.com/wasp-lang/wasp/blob/main/waspc/data/Generator/templates/server/src/config.js).
+
+## 1. Customize global middleware
+
+If you would like to modify the middleware for _all_ operations and APIs, you can do something like:
+
+```c title=todoApp.wasp
+app todoApp {
+  // ...
+
+  server: {
+    setupFn: import setup from "@server/serverSetup.js",
+    middlewareConfigFn: import { serverMiddlewareFn } from "@server/serverSetup.js"
+  },
+}
+```
+
+```ts title=src/server/serverSetup.js
+import cors from 'cors'
+import { MiddlewareConfigFn } from '@wasp/middleware'
+import config from '@wasp/config.js'
+
+export const serverMiddlewareFn: MiddlewareConfigFn = (middleware) => {
+  // Example of adding an extra domains to CORS.
+  middleware.set('cors', cors({ origin: [config.frontendUrl, 'https://example1.com', 'https://example2.com'] }))
+  return middleware
+}
+```
+
+## 2. Customize `api`-specific middleware
+
+If you would like to modify the middleware for a single API, you can do something like:
+
+```c title=todoApp.wasp
+api webhookCallback {
+  fn: import { webhookCallback } from "@server/apis.js",
+  middlewareConfigFn: import { webhookCallbackMiddlewareFn } from "@server/apis.js",
+  httpRoute: (POST, "/webhook/callback"),
+  auth: false
+}
+```
+
+```ts title=src/server/apis.ts
+import express from 'express'
+import { WebhookCallback } from '@wasp/apis/types'
+import { MiddlewareConfigFn } from '@wasp/middleware'
+
+export const webhookCallback: WebhookCallback = (req, res, _context) => {
+  res.json({ msg: req.body.length })
+}
+
+export const webhookCallbackMiddlewareFn: MiddlewareConfigFn = (middleware) => {
+  console.log('webhookCallbackMiddlewareFn: Swap express.json for express.raw')
+  
+  middleware.delete('express.json')
+  middleware.set('express.raw', express.raw({ type: '*/*' }))
+
+  return middleware
+}
+
+```
+
+:::note
+This gets installed on a per-method basis. Behind the scenes, this results in something like:
+
+```js
+router.post('/webhook/callback', webhookCallbackMiddleware, ...)
+```
+:::
+
+## 3. Customize path-path middleware
+
+If you would like to modify the middleware for all API routes under some common path, you can do something like:
+
+```c title=todoApp.wasp
+namespace fooBar {
```

Comment:
So this is cool! How are you feeling about `namespace` name? I find it pretty generic, maybe too generic. Is there a good, more specific name? That makes it clear we are dealing with `api` here? `apiRoute`? Even `apiNamespace` or `apiPath` is better than just `namespace` I think.

Reply:
+1 on `apiNamespace`, I would also think about what is this thing? For now it's a path definition which has custom middleware as one of its options. Will we expand in the future? Maybe add some sort of auth option to control access to a subset of routes? If yes, which name makes sense? 

Laravel calls this things `routeGroup` for example 🤷  https://laravel.com/docs/10.x/routing#route-groups

Reply:
@shayneczyzewski don't forget the VS Code extension update as well 
<img width="684" alt="Screenshot 2023-04-19 at 21 35 56" src="https://user-images.githubusercontent.com/2223680/233186777-35147d89-bbbf-4c07-b332-592b3d667a68.png">

Reply:
I would certainly prefix it with `api` for now. If in the future we come up with better DSL for all this stuff then fine, but for now I think it is important to make it clear that it is related to the `api`. So `apiNamespace` or `apiRoute` is ok with me.

Reply:
I'll just go with `apiNamespace` for v1. Using `apiRoute` feels like it could confuse people with our existing `route` notion. I think this should probably feel pretty distinct and vague until we know how it will evolve, and a name we don't love will force us to come back to it lol Sound ok?